Sources/SourceControl/GitRepository.swift

```diff
@@ -245,6 +258,25 @@ public class GitRepository: Repository, WorkingCheckout {
         }
     }
 
+    public func remove() throws {
+        try queue.sync {
+            // If there is a staged or unstaged diff, don't remove the
+            // tree. This won't detect new untracked files, but it is
+            // just a safety measure for now. This is only valid in a work tree.
+            let isBareRepo = try Git.runPopen([Git.tool, "-C", path.asString, "rev-parse", "--is-bare-repository"]).chomp() == "true"
```

Comment:
I think it is better to track explicitly because it isn't going to change after the repository is opened, and there are a large class of operations that only make sense on one kind of repository or the other. We could model that with subclasses, or just an internal flag, but either way I think the Repository implementation should track it.